File: katalyze/src/main/java/rules/Criteria/AllTeamsSolvedOneProblem.java
package rules.Criteria;

import rules.StandingsCriterion;
import model.Score;
import model.Standings;

public class AllTeamsSolvedOneProblem implements StandingsCriterion {
	
	public boolean isFulfilled(Standings standings) {
		for (Score score : standings) {
			if (score.solvedProblemCount() == 0) {
				return false;
			}
		}
		return true;		
	}
	
	public String message() {
		return "All teams have solved at least one problem";
	}

}

File: katalyze/src/main/kotlin/rules_kt/AllTeamsSolvedOneProblem.kt
package rules_kt

import kotlinx.coroutines.flow.flow
import model.Commentary
import model.EventImportance
import org.icpclive.cds.scoreboard.ContestStateWithScoreboard


class AllTeamsSolvedOneProblem : RuleInterface() {
    override val filters = listOf(
        FlowFilters::isAccepted,
        FlowFilters.byTrigger { contestInfo, runs ->
            runs.filter { it.isAccepted() }.map { it.teamId }.toSet()
                .containsAll(contestInfo.teams.values.filter { !it.isHidden }.map { it.id })
        },
    )

    override suspend fun process(contestStateWithScoreboard: ContestStateWithScoreboard) = flow {
        emit(
            Commentary.fromRunUpdateState(
                contestStateWithScoreboard.state, EventImportance.Breaking
            ) { _, _ -> "All teams have solved at least one problem" }
        )
    }

    override fun toString(): String {
        return "AllTeamsSolvedOneProblem()"
    }
}